util/include/Handoff.hpp

```diff
@@ -0,0 +1,94 @@
+// Concord
+//
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").  You may not use this product except in
+// compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <queue>
+#include <thread>
+#include <mutex>
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <exception>
+#include "Logger.hpp"
+
+namespace concord::util {
+/**
+ * Simple handing off to a dedicated thread functionality.
+ * Implemented by passing a function with a trivial signature void(void).
+ * Parameters are bound to an arbitrary function by a caller.
+ */
+class Handoff {
+  typedef std::lock_guard<std::mutex> guard;
+  typedef std::function<void()> func_type;
+
+ public:
+  Handoff() {
+    thread_ = std::thread([this] {
+      try {
+        for (;;) pop()();
+      } catch (ThreadCanceledException& e) {
+        LOG_DEBUG(getLogger(), "thread stopped " << std::this_thread::get_id());
+      }
+    });
+  }
+  ~Handoff() {
+    stopped_ = true;
+    queue_cond_.notify_one();
+    auto tid = thread_.get_id();
+    thread_.join();
+    LOG_DEBUG(getLogger(), "thread joined " << tid);
+  }
+
+  void push(func_type f) {
+    {
+      guard g(queue_lock_);
+      task_queue_.push(f);
+    }
+    queue_cond_.notify_one();
+  }
+
+ protected:
+  class ThreadCanceledException : public std::runtime_error {
+   public:
+    ThreadCanceledException() : std::runtime_error("thread cancelled") {}
+    const char* what() const noexcept override { return std::runtime_error::what(); }
+  };
+
+  func_type pop() {
+    while (true) {  // for spurious wake ups
+      std::unique_lock<std::mutex> ul(queue_lock_);
+      queue_cond_.wait(ul, [this] { return !(task_queue_.empty() && !stopped_); });
```

Comment:
[Minor]: Please correct me if I am wrong, but I think this comment is not entirely correct as spurious wake ups are handled in the predicate passed to `wait()`. It seems this loop allows us to exhaust the queue after `stopped_` has been set.

Reply:
correct. done